beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,305 @@
+        chain
+            .task_executor
+            .spawn_handle(
+                async move {
+                    let mut timer = metrics::start_timer_vec(
+                        &metrics::DATA_COLUMN_SIDECAR_COMPUTATION,
+                        &[&blobs_cloned.len().to_string()],
+                    );
+                    let blob_refs = blobs_cloned
+                        .iter()
+                        .filter_map(|b| b.as_ref().map(|b| &b.blob))
+                        .collect::<Vec<_>>();
+                    let data_columns_result =
+                        blobs_to_data_column_sidecars(&blob_refs, &block_cloned, &kzg, &spec)
+                            .discard_timer_on_break(&mut timer);
+                    drop(timer);
+
+                    let mut all_data_columns = match data_columns_result {
+                        Ok(d) => d,
+                        Err(e) => {
+                            error!(
+                                logger,
+                                "Failed to build data column sidecars from blobs";
+                                "error" => ?e
+                            );
+                            return;
+                        }
+                    };
+
+                    if let Err(e) = data_columns_sender.try_send(all_data_columns.clone()) {
+                        error!(logger, "Failed to send computed data columns"; "error" => ?e);
+                    };
+
+                    // Check indices from cache before sending the columns, to make sure we don't
+                    // publish components already seen on gossip.
+                    let is_supernode = chain_cloned
+                        .data_availability_checker
+                        .get_custody_columns_count()
+                        == spec.number_of_columns;
+
+                    // At the moment non supernodes are not required to publish any columns.
+                    // TODO(das): we could experiment with having full nodes publish their custodied
+                    // columns here.
+                    if !is_supernode {
+                        return;
+                    }
+
+                    // To reduce bandwidth for supernodes: permute the columns to publish and
+                    // publish them in batches. Our hope is that some columns arrive from
+                    // other supernodes in the meantime, obviating the need for us to publish
+                    // them. If no other publisher exists for a column, it will eventually get
+                    // published here.
+                    // FIXME(das): deduplicate this wrt to gossip/sync methods
+                    all_data_columns.shuffle(&mut rand::thread_rng());
+
+                    let batch_size =
+                        all_data_columns.len() / SUPERNODE_DATA_COLUMN_PUBLICATION_BATCHES;
+                    for batch in all_data_columns.chunks(batch_size) {
```

Comment:
Consider a case where the columns to publish are batched into `[[1,2,3,4],[5,6,7,8]]`. At the moment to publish the first batch, `[1,2,3,4]` are already seen, but `[5,6,7,8]`. With this design the node will sit idle for `SUPERNODE_DATA_COLUMN_PUBLICATION_BATCH_INTERVAL` and do nothing. 

Is that expected? Do we want max the rate limited bandwidth of the node, or distribute publishing even if some supernodes don't publish anything for an interval?

Reply:
It's random, so the chance of having a first batch that's already fully known is quite unlikely.